### DISCUSSION TITLE
How to adjust pyplot to screens

### DISCUSSION THREAD

**Discussion by Guilhermwn** [zauberzeug/nicegui#3277] [Q&A] [answered]
### Question

Hello to everyone, firstly thank you for such an amazing GUI library.

I'm working on an app that allows users to create plots. The user inputs the data for the plot, and I process this data using `pyplot`. However, I'm having trouble with the placement of the plot on the screen. I want to achieve the same resizing effect that Streamlit has when plotting with `pyplot`, where the plot adjusts its size dynamically to fit the space, similar to how an image would resize.

### The code i'm using to create  and update the plot is in a class
```python
def update_plot(self):
    with self.plot:
        self.ax.clear()
        
        if self.graph_type == self.graph_options[0]:
            self.ax.plot(
                self.x_axes, 
                self.y_axes, 
                color = self.color_map[self.data_color],
                linestyle = self.line_style_map[self.line_style]
                )
            
        elif self.graph_type == self.graph_options[1]:
            self.ax.scatter(
                self.x_axes, 
                self.y_axes,
                color = self.color_map[self.data_color],
                marker = self.markers_style_map[self.markers_style]
                )
        
        self.ax.set_xlabel(self.x_label)
        self.ax.set_ylabel(self.y_label)
        self.ax.set_title(self.graph_title)
    ui.update(self.plot)
```

### And this is the code calling the class and showing the plot on nicegui:
```python
with ui.element('div').classes('w-full flex flex-col items-center'): 
    with ui.element('div').classes('w-full flex flex-col items-center my-5'):
        with ui.pyplot(close=False)as plot:
            pyplot_manager.set_pyplot(plot)
        
    with ui.element('div').classes('w-full flex flex-col items-center'):     
        with ui.row():
            ui.button('Baixar JPG')
            ui.button('Baixar PDF')
```

##### This is the result when open in a wide screen browser such as a notebook or pc, it is perfect
![image](https://github.com/zauberzeug/nicegui/assets/32494940/f0176acd-a6ab-430e-b529-315dd132beda)

##### But this is the result on mobile, the plot is not resized to the screen properly
<img src="https://github.com/zauberzeug/nicegui/assets/32494940/7d08ed32-0d48-4a55-8bb5-b2f39bbaa632" width="200">

##### And this is the result i'm trying to achieve, that streamlit offers, as you can see the plot was kind of resized like an image to fit the space of the mobile screen
 <img src="https://github.com/zauberzeug/nicegui/assets/32494940/573f1408-2363-49bc-a6f9-d0918bfa0b45" width="200">

**Comment by falkoschindler** [comment]
Hi @Guilhermwn,

The plot looks fine to me, as it nicely fits the width of the mobile screen. The other elements are a bit narrow though. But I probably don't get what you're trying to achieve. 🤔

Apart from that, can you, please, try to create a minimum reproducible code example? This would allow us and the community to help more efficiently. Thanks!
[Why is it important to provide a minimal reproducible example?](https://github.com/zauberzeug/nicegui/wiki/FAQs#why-is-it-important-to-provide-a-minimal-reproducible-example)

**Reply by Guilhermwn** [reply]
Hello @falkoschindler, sorry for the slow response.

this is simple code i created to show:

```python
@ui.page('/testes')
def testes():
    # enabling dark mode to better view
    ui.dark_mode().enable()

    # custom component using ui.header
    header_layout()
    
    # using column in full width to display all content
    with ui.column().classes('w-full bg-red-200'):
        ui.markdown('# Space reserved for content').classes('text-black')
        
        # creating a div for the content of the matplotlib plot
        with ui.element('div').classes('w-full bg-green-200'):
            ui.markdown('### Original matplotlib plot inserted using ui.pyplot()').classes('text-black')

            # matplotlib plot initiated
            with ui.pyplot(close=False) as plot:
                ax = plot.fig.add_subplot()
                ax.plot(
                    [1,2,3,4], 
                    [1,2,3,4], 
                    color = 'green',
                    linestyle = '-.'
                    )
```
### PC View
In the wide view using a PC, everything looks fine. The red border represents the column container, and the green border is the div for the plot container. There is sufficient space for the plot within the designated area.
![image](https://github.com/zauberzeug/nicegui/assets/32494940/b91df775-39ee-4620-8c2f-47db1d5a0158)

### Mobile View
However, on mobile devices, the plot does not respect the container's width. The following image shows the default zoom on mobile, with no zooming in or out applied. The plot appears to ignore the container's constraints, causing layout issues.
<img src="https://github.com/zauberzeug/nicegui/assets/32494940/6c42f5fe-def2-4075-8cbb-2368848997b7" width="300">

### Minimal Zoom on Mobile
This screenshot shows the plot with the minimal zoom applied on mobile. The plot still exceeds the container's width.
<img src="https://github.com/zauberzeug/nicegui/assets/32494940/aede6d13-ca89-4dae-b3a8-65b7649eaa33" width="300">

### Summary
The plot generated by ui.pyplot() appears to disregard the width constraints of its container on mobile devices, leading to an oversized display that disrupts the overall layout. I would appreciate any guidance or solutions to ensure the plot fits within the container on mobile devices.

**Comment by python-and-novella** [comment]
Before `ui.pyplot` changed,you could achieve it by other way like this:
```python3
from nicegui import ui
from matplotlib import pyplot
from lxml import etree
@ui.page('/')
def index():
    with ui.column().classes('w-full bg-red-200'):
        ui.markdown('# Space reserved for content').classes('text-black')
        with ui.element('div').classes('w-full bg-green-200'):
            ui.markdown('### Original matplotlib plot inserted using ui.pyplot()').classes('text-black')
            with ui.pyplot(close=False) as plot1:
                ax = plot1.fig.add_subplot()
                ax.plot(
                    [1,2,3,4], 
                    [1,2,3,4], 
                    color = 'green',
                    linestyle = '-.'
                    )
            def set_svg_size(plot: ui.pyplot=None, width :str='100%', height: str='100%') -> None:
                if plot is None:
                    pass
                else:
                    e = etree.fromstring(plot._props['innerHTML'].encode())
                    e.set('width', width)
                    e.set('height', height)
                    plot._props['innerHTML']=etree.tostring(e).decode()
                    plot.update()
            # way 1:
            # set svg's width and height to 100%.
            set_svg_size(plot1)
            # way 2:
            # save as file,and display it by imgage.
            # Before image,should hide pyplot.
            # Here just shows two ways,without hiding.
            plot1.fig.savefig('./plot.svg')
            #plot1.set_visibility(False)
            ui.image('./plot.svg')
ui.run()
```

**Comment by falkoschindler** [comment]
Ah, now I understand! The question is, how to automatically shrink `ui.pyplot` to its container width. Minimum reproduction:
```py
with ui.card().classes('w-[400px]'):
    with ui.pyplot() as plot:
        plot.fig.gca().plot()
```
![Screenshot 2024-06-27 at 08 13 48](https://github.com/zauberzeug/nicegui/assets/5767091/06e0cd95-ab49-4657-85a3-28720b1850db)

**Reply by Guilhermwn** [reply]
Thank you very much. In my next questions, I will try to make the question clearer and provide a better MRE

**Comment by falkoschindler** [comment, marked as answer]
Here's a workaround adding "width:100%;height:100%" to the inner HTML:
```py
with ui.card().classes('w-[400px]'):
    with ui.pyplot() as plot:
        plot.fig.gca().plot()
    plot._props['innerHTML'] = plot._props['innerHTML'].replace('<svg', '<svg style="width:100%;height:100%"')
    plot.update()
```

But I think we can add these attributes by default so that the SVG automatically fills the size of the `ui.pyplot` element.

**Reply by falkoschindler** [reply]
I created PR #3289 to automatically scale the SVG to the size of the `ui.pyplot` elemement.

**Reply by Guilhermwn** [reply]
Thank you very much for the reply and for the solution. I added it to my code and it worked fine on all width's i tried 👌

**Reply by Guilhermwn** [reply]
I'll explain how i used this solution you gave me, on my use case. Since I created the class `PyplotManager` to manage the updating values of the plot, I needed to add the `props` every time I updated an aspect of the plot.

### Main Code:
In the main code, I set up the plot and managed its properties. Here, I included the `props` adjustment during the initial creation of the plot.

```python
pyplot_manager = PyplotManager()

with ui.element('div').classes('w-full my-5'):
    with ui.pyplot(close=False) as plot:
        pyplot_manager.set_pyplot(plot)
        pyplot_manager.download_plot()
    
    plot._props['innerHTML'] = plot._props['innerHTML'].replace('<svg', '<svg style="width:100%;height:100%"')
    plot.update()
```

### Class Code:
In the class code, which handles the updates to the `plot`, I also needed to adjust the props parameter every time an update occurred. My website, using NiceGUI, receives parameters like **x_axes**, **y_axes,** and titles with input data from the user. I use these parameters to add and update the plot, essentially creating a simple plot creator using NiceGUI.

When I updated the plot without using the `props`, the width set in the main code would revert to its original size and escape the container. To maintain the plot size within the parent container, I had to add the `props` adjustment at the end of the class method for updates.

```python
def update_plot(self):
    with self.plot:
        self.ax.clear()
        
        if self.graph_type == self.graph_options[0]:
            self.ax.plot(
                self.x_axes, 
                self.y_axes, 
                color=self.color_map[self.data_color],
                linestyle=self.line_style_map[self.line_style]
            )
        
        elif self.graph_type == self.graph_options[1]:
            self.ax.scatter(
                self.x_axes, 
                self.y_axes,
                color=self.color_map[self.data_color],
                marker=self.markers_style_map[self.markers_style]
            )
        
        self.ax.set_xlabel(self.x_label, fontsize=14)
        self.ax.set_ylabel(self.y_label, fontsize=14)
        self.ax.set_title(self.graph_title, fontsize=14)

    self.plot._props['innerHTML'] = self.plot._props['innerHTML'].replace('<svg', '<svg style="width:100%;height:100%"')
    self.plot.update()
```


Again, thank you very much for the solution and reply.

**Reply by falkoschindler** [reply]
I'm glad we found a solution. The next release including PR #3289 should make this workaround obsolete. 🙂

**Reply by moi90** [reply]
@falkoschindler Would it make sense to apply the same to the newer `ui.matplotlib`?

**Reply by falkoschindler** [reply]
Great idea, ~~@Guilhermwn~~ @moi90! I wasn't aware of this bug. Minimum reproduction:
```py
with ui.card().classes('w-80'):
    with ui.matplotlib().figure as fig:
        fig.gca().plot([1, 2, 3], [1, 4, 3])
```
![Screenshot 2025-04-08 at 15 09 15](https://github.com/user-attachments/assets/8517b8b0-5a59-4a2d-9875-e3e0a9dbbd2f)

Would you like to create a pull request similar to PR #3289? 🙂

**Reply by evnchn** [reply]
Wait a moment, please. 

Please don't straight-up follow the implementation in #3289. 

Checking #3843, use this implementation instead: https://github.com/zauberzeug/nicegui/commit/70081770b49b2cfc92e503196dcb715c195d964a

**Reply by falkoschindler** [reply]
Great catch, @evnchn! Using `default_classes` is definitely the right way to add - well - default classes. 👍🏻

**Reply by moi90** [reply]
@falkoschindler I'm assuming that you wanted to `@ *me* instead of Guilhermwn.

But I see that @evnchn already worked on this: https://github.com/zauberzeug/nicegui/pull/4595

Thanks!

**Reply by falkoschindler** [reply]
@moi90 Oh, sorry, I accidentally picked the wrong GitHub handle.
Yes, the fix has already been implemented and released in version 2.14.1. 🏎️
